three_pass.go

```diff
@@ -181,25 +221,39 @@ func (jp *ThreePassJpake[P, S]) computeZKP(x S, generator P, y P) (ZKPMsg[P, S],
 	if err != nil {
 		return ZKPMsg[P, S]{}, err
 	}
-	cS, err := jp.curve.NewScalar().SetBigInt(c)
-	if err != nil {
-		return ZKPMsg[P, S]{}, err
-	}
 	return ZKPMsg[P, S]{
 		T: t,
 		R: rS,
-		C: cS,
 	}, err
 }
 
 func (jp *ThreePassJpake[P, S]) checkZKP(msgObj ZKPMsg[P, S], generator, y P) bool {
-	chal := generator.Bytes()
-	chal = append(chal, msgObj.T.Bytes()[:]...)
-	chal = append(chal, y.Bytes()[:]...)
+	chal := []byte{}
+	binary.BigEndian.AppendUint32(chal, uint32(len(generator.Bytes())))
+	chal = append(chal, generator.Bytes()...)
+	binary.BigEndian.AppendUint32(chal, uint32(len(msgObj.T.Bytes())))
+	chal = append(chal, msgObj.T.Bytes()...)
+	binary.BigEndian.AppendUint32(chal, uint32(len(y.Bytes())))
+	chal = append(chal, y.Bytes()...)
+	binary.BigEndian.AppendUint32(chal, uint32(len(jp.OtherUserID)))
 	chal = append(chal, jp.OtherUserID...)
-	c := (new(big.Int).SetBytes(jp.hashFn(chal)))
+
+	c := (new(big.Int).SetBytes(jp.config.hashFn(chal)))
 	c = c.Mod(c, jp.curve.Params().N)
-	// TODO: ensure c is not 0 (i think)
+
+	// if c is zero
+	if c.BitLen() == 0 {
+		return false
+	}
+	// validate T is not 0
+	if msgObj.T.Equal(jp.curve.NewPoint()) == 1 {
+		return false
+	}
+
+	// validate public key is not at infinity
+	if msgObj.R.BigInt().BitLen() == 0 {
```

Comment:
Testing at least suggests that here could be a problem when R is not checked if being the curve order N